src/main/java/uk/gov/ons/census/casesvc/messaging/SampleReceiver.java

```diff
@@ -63,52 +66,92 @@ public SampleReceiver(
   @Transactional
   @ServiceActivator(inputChannel = "caseSampleInputChannel")
   public void receiveMessage(CreateCaseSample createCaseSample) {
-
-    Case caze = persistToDatabase(createCaseSample);
-
-    CaseCreatedEvent caseCreatedEvent = prepareEventToEmit(caze);
-
-    rabbitTemplate.convertAndSend(emitCaseEventExchange, "", caseCreatedEvent);
+    Case caze = saveCase(createCaseSample);
+    UacQidLink uacQidLink = saveUacQidLink(caze);
+    emitUacUpdatedEvent(uacQidLink, caze);
+    emitCaseCreatedEvent(caze);
+    logEvent(uacQidLink, CASE_CREATED_EVENT_DESCRIPTION);
+    logEvent(uacQidLink, UAC_QID_LINKED_EVENT_DESCRIPTION);
   }
 
-  private Case persistToDatabase(CreateCaseSample createCaseSample) {
-
+  private Case saveCase(CreateCaseSample createCaseSample) {
     Case caze = mapperFacade.map(createCaseSample, Case.class);
     caze.setCaseId(UUID.randomUUID());
     caze.setState(CaseState.ACTIONABLE);
     caze = caseRepository.saveAndFlush(caze);
+    return caze;
+  }
 
+  private UacQidLink saveUacQidLink(Case caze) {
     UacQidLink uacQidLink = new UacQidLink();
     uacQidLink.setId(UUID.randomUUID());
     uacQidLink.setUac(iacDispenser.getIacCode());
     uacQidLink.setCaze(caze);
     uacQidLink = uacQidLinkRepository.saveAndFlush(uacQidLink);
 
-    String qid =
-        qidCreator.createQid(createCaseSample.getTreatmentCode(), uacQidLink.getUniqueNumber());
+    // The unique number has been generated by the DB by this point, so we can use it
+    String qid = qidCreator.createQid(caze.getTreatmentCode(), uacQidLink.getUniqueNumber());
     uacQidLink.setQid(qid);
     uacQidLinkRepository.save(uacQidLink);
 
+    return uacQidLink;
+  }
+
+  private void logEvent(UacQidLink uacQidLink, String eventDescription) {
     uk.gov.ons.census.casesvc.model.entity.Event loggedEvent =
         new uk.gov.ons.census.casesvc.model.entity.Event();
     loggedEvent.setId(UUID.randomUUID());
     loggedEvent.setEventDate(new Date());
-    loggedEvent.setEventDescription(EVENT_DESCRIPTION);
+    loggedEvent.setEventDescription(eventDescription);
     loggedEvent.setUacQidLink(uacQidLink);
     eventRepository.save(loggedEvent);
+  }
 
-    return caze;
+  private void emitUacUpdatedEvent(UacQidLink uacQidLink, Case caze) {
+    Event event = createEvent(EventType.UAC_UPDATED);
+
+    Uac uac = new Uac();
+    uac.setActive(true);
+    uac.setCaseId(caze.getCaseId().toString());
+    uac.setCaseType(caze.getAddressType());
+    uac.setCollectionExerciseId(caze.getCollectionExerciseId());
+    uac.setQuestionnaireId(uacQidLink.getQid());
+    uac.setUacHash(Sha256Helper.hash(uacQidLink.getUac()));
+    uac.setUac(uacQidLink.getUac());
```

Comment:
I might be wrong(!), but are we additionally sending the actual UAC in the `UACUpdated` event?

Reply:
Just re-read the Trello comments - all good 👍